### PR TITLE
KEYCLOAK-14767 OpenShift Review Endpoint audience fix

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.redhat.com/browse/KEYCLOAK-14767

This Pull Request replaces https://github.com/keycloak/keycloak/pull/7268. It introduces a **mandatory** audience check for OpenShift Token Review Endpoint.

### CODE REVIEW

**Review by pedroigor — APPROVED** [review]
@slaskawi Ready to merge ?

**Comment by slaskawi** [comment]
@pedroigor I'd like to wait for @vmuzikar or @stianst to review it. Thanks for approving BTW!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java [hunk 1]
```diff
@@ -228,29 +241,10 @@ public void customScopes() {
     }
 
     @Test
-    public void emptyScope() throws Exception {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java:231]: Why to remove this test? Is it not valid anymore?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java [hunk 2]
```diff
@@ -228,29 +241,10 @@ public void customScopes() {
...
-            } finally {
-                oauth.openid(true);
-            }
-        }
-        // The default client scopes should be same like before.
-        int scopesAfterSize = testRealm().clients().get(clientRep.getId()).getDefaultClientScopes().size();
-        assertEquals(scopesBefore.size(), scopesAfterSize);
+    public void emptyAudience() {
+        new Review().username("empty-audience")
+                .invoke()
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java:246]: Is this really invoked without audience? I can see it added here:
https://github.com/keycloak/keycloak/blob/39b05b499c238899bb9d02ef681e500dc58fd38e/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java#L379
But maybe I'm just missing something. :)

File: core/src/main/java/org/keycloak/TokenVerifier.java [hunk 3]
```diff
@@ -366,11 +366,18 @@ private void removeCheck(Predicate<? super T> check) {
...
      *
-     * @param expectedAudience Audience, which needs to be in the target token. Can't be null
+     * @param expectedAudiences Audiences, which needs to be in the target token. Can be <code>null</code>.
      * @return This token verifier
      */
-    public TokenVerifier<T> audience(String expectedAudience) {
-        return this.replaceCheck(AudienceCheck.class, true, new AudienceCheck(expectedAudience));
+    public TokenVerifier<T> audience(String... expectedAudiences) {
+        if (expectedAudiences == null || expectedAudiences.length == 0) {
+            return this.replaceCheck(AudienceCheck.class, true, new AudienceCheck(null));
```
Comment (vmuzikar) [inline core/src/main/java/org/keycloak/TokenVerifier.java:374]: Just a silly question. For some reason, originally it was prohibited for `expectedAudience` to be null. Why to allow it now? And are we 100% sure this doesn't create e.g. some vulnerability?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java (hunk 1 above)
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java:231]: Yes, this is not valid anymore. If you remove all the scopes, you won't get any audience in the token. So with adding audience to the Token Retrospection Endpoint, this has no sense anymore.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java (hunk 2 above)
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/openshift/OpenShiftTokenReviewEndpointTest.java:246]: It's the other way around. The `account` audience is being added to the **expected** audience. The Token itself will have empty audience emitted from the Token Endpoint.

File: core/src/main/java/org/keycloak/TokenVerifier.java (hunk 3 above)
Comment (slaskawi) [reply core/src/main/java/org/keycloak/TokenVerifier.java:374]: The behavior remains the same here. In the old code, (with `String expectedAudience`), if you pass null there, you will effectively invoke `return this.replaceCheck(AudienceCheck.class, true, new AudienceCheck(null));`

The problem with the updated code is that the argument can be a null or an empty array. In both cases, I'm explicitly adding `return this.replaceCheck(AudienceCheck.class, true, new AudienceCheck(null));`.

In both cases, this blows up in the runtime [here](https://github.com/keycloak/keycloak/blob/39b05b499c238899bb9d02ef681e500dc58fd38e/core/src/main/java/org/keycloak/TokenVerifier.java#L148).

**Review by vmuzikar — APPROVED** [review]
@slaskawi Thanks for the clarification! Looks good, makes sense to me now. 👍

**Review by abstractj — APPROVED** [review]

**Pull request merged**